// vuepress-theme-hope@2.0.0-beta.210/lib/presets/BingHeroBackground.js
import { ClientOnly, usePageLang } from "@vuepress/client";
import { useStorage } from "@vueuse/core";
import { computed, defineComponent, h, onMounted } from "vue";
import "./bing-hero-background.scss";
const bingStorage = useStorage("bing-image", {
    index: 0,
    data: [],
});
export default defineComponent({
    name: "BingHeroBackground",
    // TODO: Add download button, image description and copyright information
    // props: {
    //   download: Boolean,
    // },
    setup() {
        const lang = usePageLang();
        const currentWallpaper = computed(() => {
            const info = bingStorage.value.data[bingStorage.value.index];
            const langCode = lang.value.toLowerCase().split("-").shift();
            if (info) {
                const { url, wallpaper, downloadable, locales } = info;
                return {
                    url,
                    wallpaper,
                    downloadable,
                    ...(locales[langCode] ?? locales["en"]),
                };
            }
            return null;
        });
        const getImage = () => {
            return fetch("https://bing-wallpaper.vuejs.press/api/wallpaper").then((response) => response.json());
        };
        const prev = () => {
            bingStorage.value.index--;
        };
        const next = () => {
            bingStorage.value.index++;
        };
        onMounted(() => {
            void getImage().then((res) => {
                bingStorage.value.data = res;
            });
        });
        return () => h(ClientOnly, () => currentWallpaper.value
            ? [
                h("div", {
                    class: "mask",
                    style: {
                        background: `url(${currentWallpaper.value.url}) center/cover no-repeat`,
                    },
                }),
                h("div", { class: "bing-switch" }, [
                    h("a", {
                        class: "bing-info",
                        href: currentWallpaper.value.backstage,
                        target: "_blank",
                    }, [
                        h("span", { class: "bing-location-icon" }),
                        currentWallpaper.value.title,
                    ]),
                    h("button", {
                        class: "bing-switch-left",
                        disabled: bingStorage.value.index === 0,
                        onClick: () => prev(),
                    }),
                    h("button", {
                        class: "bing-switch-right",
                        disabled: bingStorage.value.index ===
                            bingStorage.value.data.length - 1,
                        onClick: () => next(),
                    }),
                ]),
            ]
            : null);
    },
});
