// @maverick-js/signals@5.10.5/dist/prod/signals.js
import { createComputation, read, write, isFunction, update, dispose, onDispose } from './core.js';
import './symbols.js';

function signal(initialValue, options) {
  const node = createComputation(initialValue, null, options), signal2 = read.bind(node);
  signal2.set = write.bind(node);
  return signal2;
}
function isReadSignal(fn) {
  return isFunction(fn);
}
function computed(compute, options) {
  return read.bind(
    createComputation(
      options?.initial,
      compute,
      options
    )
  );
}
function effect(effect2, options) {
  const signal2 = createComputation(
    null,
    function runEffect() {
      let effectResult = effect2();
      isFunction(effectResult) && onDispose(effectResult);
      return null;
    },
    void 0
  );
  signal2.$e = true;
  update(signal2);
  return dispose.bind(signal2, true);
}
function readonly(signal2) {
  return () => signal2();
}
function isWriteSignal(fn) {
  return isReadSignal(fn) && "set" in fn;
}

export { computed, effect, isReadSignal, isWriteSignal, readonly, signal };
